src/libraries/System.Net.Http/src/System/Net/Http/Headers/HttpHeaders.cs

```diff
@@ -196,6 +205,63 @@ internal bool TryAddWithoutValidation(HeaderDescriptor descriptor, IEnumerable<s
             return true;
         }
 
+        private bool SpecializedTryAddWithoutValidation(HeaderDescriptor descriptor, string?[]? valuesArray, IList<string?>? valuesList)
+        {
+            Debug.Assert(valuesArray != null || valuesList != null);
+            int count = valuesArray != null ? valuesArray.Length : valuesList!.Count;
+
+            if (count == 0)
+            {
+                return true;
+            }
+
+            ref object? storeValueRef = ref GetValueRefOrAddDefault(descriptor);
+            object? storeValue = storeValueRef;
+
+            if (storeValue is not null || count > 1)
+            {
+                if (storeValue is not HeaderStoreItemInfo info)
+                {
+                    storeValueRef = info = new HeaderStoreItemInfo { RawValue = storeValue };
+                }
+
+                object? rawValue = info.RawValue;
+                if (rawValue is not List<string> rawValues)
+                {
+                    info.RawValue = rawValues = new List<string>();
+                    if (rawValue is not null)
+                    {
+                        rawValues.EnsureCapacity(count + 1);
+                        rawValues.Add((string)rawValue);
+                    }
+                }
```

Comment:
I can decipher the code, but it would be faster for me if there were comments explaining the individual cases we are trying to handle (I assume RawValue can be either `string` or `List<string>` (and maybe something else?)